RearrangingCars.java

```diff
@@ -0,0 +1,207 @@
+     */
+    private static  ArrayList<Step> rearrangingStepsHelper(int [] current, int [] expected){
+        
+        ArrayList<Step> steps = new ArrayList<Step>();
+        
+        //Create an inverse-mapping of the "current" rearrangement.
+        int [] inverseMapping_current = getInverseMapping(current);            
+        
+        //After obtaining the inverse-mapping, find the index corresponding to 0 (no car).
+        int empty_slot = findEmptySlot(inverseMapping_current);
+        
+        //Obtain the set of spaces (indices) which do not have correctly placed cars (and are yet to be rearranged)
+        HashSet<Integer> yetToRearrange = indices_needToRearrange(current, expected);
+        
+        //Until no space (index) has incorrectly placed car (no index needs further rearrangement)
+        while(!yetToRearrange.isEmpty()){
+            //Let "carToBeRearranged" be the car to be correctly placed in the empty slot
+            int carToBeRearranged = expected[empty_slot];                
+
+            //Obtain the index where "carToBeRearranged" is currently placed, and call it "move_from"
+            int move_from = inverseMapping_current[carToBeRearranged];
+            
+            //If "move_from" is same as the "empty_slot", then, the "expected-rearrangement" has empty slot at the same position
+            if(move_from == empty_slot){
+                //Since the empty_slot position matches with the expected rearrangement, remove it from the set of "yetToRearrange" indices
+                yetToRearrange.remove(empty_slot);
+                
+                //If there is no more rearrangement required, then break
+                if(yetToRearrange.isEmpty()){
+                    break;
+                }
+                
+                //Now, prepare to move a car from any location that has an incorrectly placed car.
+                move_from = yetToRearrange.iterator().next();     
+                carToBeRearranged = current[move_from];
+            }                
+
+            //Move an incorrectly placed car to its correct index  
+            steps.add(new Step(move_from, empty_slot));
+            
+            //Update the current rearrangement                
+            current[empty_slot] = current[move_from];
+            current[move_from] = 0;  
+            
+            //Update the inverse mapping accordingly
+            inverseMapping_current[carToBeRearranged] = empty_slot;
+            inverseMapping_current[0] = move_from;                               
+
+            //Remove the empty_slot from the set of "yetToRearrange" indices, since it is occupied now
+            yetToRearrange.remove(empty_slot);
+            
+            //The index from which the car has moved, now contains 0 (no car) and becomes the new empty_slot 
+            empty_slot = move_from;                
+        }
+        return steps;
+    }

```

Comment:
Adding comment to every line is really not necessary. For example `//If there is no more rearrangement required, then break` is a candidate to be removed.